pkg/resource/deploy/plan_executor.go

```diff
@@ -303,39 +312,54 @@ func (pe *planExecutor) refresh(callerCtx context.Context, opts Options, preview
-	// resource's position in the resource list. For example, consider the following list of resources, where each
-	// resource is a (URN, ID, Dependencies) tuple:
+	// The process of repairing dependency lists is a bit subtle. Because multiple physical
+	// resources may share a URN, the ability of a particular URN to be referenced in a dependency
+	// list can change based on the dependent resource's position in the resource list. For example,
+	// consider the following list of resources, where each resource is a (URN, ID, Dependencies)
+	// tuple:
 	//
 	//     [ (A, 0, []), (B, 0, [A]), (A, 1, []), (A, 2, []), (C, 0, [A]) ]
 	//
-	// Let `(A, 0, [])` and `(A, 2, [])` be deleted by the refresh. This produces the following intermediate list
-	// before dependency lists are repaired:
+	// Let `(A, 0, [])` and `(A, 2, [])` be deleted by the refresh. This produces the following
+	// intermediate list before dependency lists are repaired:
 	//
 	//     [ (B, 0, [A]), (A, 1, []), (C, 0, [A]) ]
 	//
-	// In order to repair the dependency lists, we iterate over the intermediate resource list, keeping track of which
-	// URNs refer to at least one physical resource at each point in the list, and remove any dependencies that refer
-	// to URNs that do not refer to any physical resources. This process produces the following final list:
+	// In order to repair the dependency lists, we iterate over the intermediate resource list,
+	// keeping track of which URNs refer to at least one physical resource at each point in the
+	// list, and remove any dependencies that refer to URNs that do not refer to any physical
+	// resources. This process produces the following final list:
 	//
 	//     [ (B, 0, []), (A, 1, []), (C, 0, [A]) ]
 	//
-	// Note that the correctness of this process depends on the fact that the list of resources is a topological sort
-	// of its corresponding dependency graph, so a resource always appears in the list after any resources on which it
-	// may depend.
-	resources := make([]*resource.State, 0, len(prev.Resources))
+	// Note that the correctness of this process depends on the fact that the list of resources is a
+	// topological sort of its corresponding dependency graph, so a resource always appears in the
+	// list after any resources on which it may depend.
+	resources := []*resource.State{}
 	referenceable := make(map[resource.URN]bool)
 	olds := make(map[resource.URN]*resource.State)
-	for _, s := range steps {
-		new := s.New()
+	for _, s := range initialResources {
+		var old, new *resource.State
+		if step, has := resourceToStep[s]; has {
+			// We produces a refresh step for this specific resource.  Use the new information about
+			// its dependencies during the update.
+			old = step.Old()
+			new = step.New()
+		} else {
+			// We didn't do anything with this resource.  However, we still may want to update its
+			// dependencies.  So use this resource itself as the 'new' one to update.
+			old = s
+			new = s
+		}
```

Comment:
this is hte part i'm the least confident about @pgavlin .  I also attempted to update the algorithm to add 'Same' steps at line 300.  However, it was unclear to me how to create a new Same-Step.  Specifically: NewSameStep needs to be passed a RegisterResourceEvent, and it wasn't clear to me how i would get such a value passed in.  Thoughts?

Reply:
> Specifically: NewSameStep needs to be passed a RegisterResourceEvent, and it wasn't clear to me how i would get such a value passed in. Thoughts?

All you need is a value that satisfies the `RegisterResourceEvent` interface. You could do something like:
```go
type registerSameEvent struct {
    goal *resource.Goal
}

func newRegisterSameEvent(state *resource.State) RegisterResourceEvent {
    goal := // create a new goal using the values in `state`
    return &registerSameEvent{goal}
}

func (e registerSameEvent) event() {}

func (e registerSameEvent) Goal() *resource.Goal {
    return e.goal
}

func (e registerSameEvent) Done(_ *RegisterResult) {}
```

Reply:
(that said, I think what you have here is fine)

Reply:
ok.  if the existing approach is fine, def sign off and i'll merge in.  thanks!